x/distribution/types/delegator_info.go

```diff
@@ -42,16 +50,49 @@ func (di DelegationDistInfo) WithdrawRewards(wc WithdrawContext, vi ValidatorDis
 	vi = vi.UpdateTotalDelAccum(wc.Height, totalDelShares)
 
 	if vi.DelAccum.Accum.IsZero() {
```

Comment:
I don't understand why this code block is necessary.  It's a slight optimization maybe, but it increases surface area for bugs.  Also, if vi.DelAccum.Accum.IsZero(), di.DelPoolWithdrawHeight I don't think would matter anyways (even though it would be correct to update the height here).

Reply:
hmmm - yeah maybe this is an “over-optimization” and we should just delete this whole block instead of adding this defensive check

Reply:
`I’m not saying we should add the panic after it was clearly updated. there’s a timeline of dependencies here… withdrawal <-------> further-add-to-delegation-shares currently we made the bug fix where we update the withdrawal height on the left hand side. we want the very far right hand side, which is not the highest level of code… it’s a bit deeper in the call stack.`

Reply:
Refining my thinking...

I was trying to refer to (I think) in x/stake/keeper/distribution.go,

```func (k Keeper) Delegate(ctx sdk.Context, delAddr sdk.AccAddress, bondAmt sdk.Coin,
	validator types.Validator, subtractAccount bool) (newShares sdk.Dec, err sdk.Error) {
...
	// Get or create the delegator delegation
	delegation, found := k.GetDelegation(ctx, delAddr, validator.OperatorAddr)
...
	// call the appropriate hook if present
	if found {
		k.OnDelegationSharesModified(ctx, delAddr, validator.OperatorAddr)
	} else {
		k.OnDelegationCreated(ctx, delAddr, validator.OperatorAddr)
	}
...
	// Update delegation
	delegation.Shares = delegation.Shares.Add(newShares)
	delegation.Height = ctx.BlockHeight()
	k.SetDelegation(ctx, delegation)
}
```

Was the problem that we were calling OnDelegateSharesModified here, but because the shares was zero, we didn't update the height?

Some thoughts... OnDelegatorSharesModified here should be called WillModifyDelegatorShares maybe, otherwise it appears to be a misnomer.  WillCreateDelegation as well... since the delegation hasn't been created or modified yet.  Or, maybe we can keep the old info and pass it into OnDelegationSharesModified.  Not a big deal ATM, just pointing out an inconsistency.

If we had an DidModifyDelegationShares, we could add the check in the distribution keeper to check that the withdrawal height was modified.  But also, maybe this isn't necessary.

Reply:
Agreed - ref https://github.com/cosmos/cosmos-sdk/issues/2538